src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -535,26 +602,39 @@ private void InheritNullableStateOfFieldOrProperty(int targetContainerSlot, int
             }
         }
 
-        private void InheritNullableStateOfAnonymousTypeInstance(TypeSymbol targetType, int targetSlot, int valueSlot, bool isByRefTarget)
+        private void InheritDefaultState(int targetSlot)
         {
             Debug.Assert(targetSlot > 0);
-            Debug.Assert(targetType.IsAnonymousType && targetType.IsClassType());
 
-            foreach (var member in targetType.GetMembersUnordered())
+            // Reset the state of any members of the target.
+            for (int slot = targetSlot + 1; slot < nextVariableSlot; slot++)
             {
-                if (member.Kind != SymbolKind.Property)
+                var variable = variableBySlot[slot];
+                if (variable.ContainingSlot != targetSlot)
                 {
                     continue;
                 }
+                this.State[slot] = !variable.Symbol.GetTypeOrReturnType().IsNullable;
+                InheritDefaultState(slot);
+            }
+        }
 
-                var propertySymbol = (PropertySymbol)member;
+        private void InheritNullableStateOfTrackableType(int targetSlot, int valueSlot, bool isByRefTarget)
```

Comment:
>InheritNullableStateOfTrackableType [](start = 21, length = 35)

It feels like this function should do something about target's children that do not match any slot under valueSlot. It would be more robust and clear to not assume that InheritDefaultState was called before this function.